lib/models/page.js

```diff
@@ -830,84 +906,122 @@ module.exports = function(crowi) {
-          Page.pushRevision(newPage, newRevision, user)
-            .then(function(data) {
-              resolve(data)
-              pageEvent.emit('create', data, user)
-            })
-            .catch(function(err) {
-              debug('Push Revision Error on create page', err)
-              return reject(err)
-            })
-        })
-      })
-    })
+    const savedPage = await newPage.save()
+
+    const newRevision = Revision.prepareRevision(newPage, body, user, { format: format })
+    try {
+      const data = await Page.pushRevision(savedPage, newRevision, user)
+      pageEvent.emit('create', data, user)
+      return data
+    } catch (err) {
+      debug('Push Revision Error on create page', err)
+      throw err
+    }
   }
 
-  pageSchema.statics.updatePage = function(pageData, body, user, options = {}) {
+  pageSchema.statics.updatePage = async function(pageData, body, user, options = {}) {
     const Page = this
     const Revision = crowi.model('Revision')
     const Bookmark = crowi.model('Bookmark')
     const grant = options.grant || null
+    const bodyConfirmed = options.bodyConfirmed
+
+    const expirationAt = await Page.calculationRevisionExpirationAt({
+      revision: pageData.populated('revision') ? pageData.revision : null,
+      lifetime: pageData.lifetime,
+      bodyConfirmed,
+      pageId: pageData._id,
+      userId: user._id,
+    })
+
     // update existing page
-    const newRevision = Revision.prepareRevision(pageData, body, user)
+    const newRevision = Revision.prepareRevision(pageData, body, user, { expirationAt })
 
-    return new Promise(async (resolve, reject) => {
-      Page.pushRevision(pageData, newRevision, user)
-        .then(async revision => {
-          const bookmarkCount = await Bookmark.countByPageId(pageData._id)
+    try {
+      const savedPage = await Page.pushRevision(pageData, newRevision, user)
 
-          if (grant != pageData.grant) {
-            return Page.updateGrant(pageData, grant, user).then(function(data) {
-              resolve(data)
-              pageEvent.emit('update', data, user, bookmarkCount)
-            })
-          } else {
-            resolve(pageData)
-            pageEvent.emit('update', pageData, user, bookmarkCount)
-          }
-        })
-        .catch(function(err) {
-          debug('Error on update', err)
-          debug('Error on update', err.stack)
-        })
-    })
+      const bookmarkCount = await Bookmark.countByPageId(pageData._id)
+
+      // If need to update grant
+      if (grant != pageData.grant) {
+        const p = await Page.updateGrant(savedPage, grant, user)
+        pageEvent.emit('update', p, user, bookmarkCount)
+        return p
+      }
+
+      pageEvent.emit('update', savedPage, user, bookmarkCount)
+      return savedPage
+    } catch (err) {
+      debug('Error on update', err)
+      debug('Error on update', err.stack)
+      throw err
```

Comment:
In previous impl, users can't notice errors that occured at updating. So, I change behavior on this methods at first step.